### PR TITLE
Ginkgo: Add a test to check cilium image updates

### PR DESCRIPTION
**Pull request by eloycoto**
This test check that if we update from stable to master all keep working
and no problems with data migrations.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/Nightly.go [hunk 1]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
 		return appPods
 	}
 
-	It("Check Kubernetes Example is working correctly", func() {
-		var path = "../examples/kubernetes/cilium.yaml"
-		var result bytes.Buffer
-		newCiliumDSName := fmt.Sprintf("cilium_ds_%s.json", helpers.MakeUID())
+	Context("Cilium DS from example", func() {
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Instead of DS, write out DaemonSet.

File: test/k8sT/Nightly.go [hunk 2]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
 
-		fp, err := os.Create(newCiliumDSName)
-		defer fp.Close()
-		Expect(err).To(BeNil())
+		waitEndpointReady := func() {
+			ciliumPods, err := kubectl.GetCiliumPods(helpers.KubeSystemNamespace)
+			Expect(err).To(BeNil(), "cannot retrieve cilium pods")
+			for _, pod := range ciliumPods {
+				ExpectWithOffset(1, kubectl.CiliumEndpointWait(pod)).To(BeTrue(),
+					"Pods are not ready in pod %v", pod)
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Change message to "Pod %s is not ready"

File: test/k8sT/Nightly.go [hunk 3]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
+			defer fp.Close()
+			Expect(err).To(BeNil())
+
+			fmt.Fprint(fp, result.String())
+
+			kubectl.Apply(helpers.GetFilePath(newCiliumDSName)).ExpectSuccess(
+				"cannot apply cilium example daemonset")
+
+			status, err := kubectl.WaitforPods(
+				helpers.KubeSystemNamespace, "-l k8s-app=cilium", 300)
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Use time.Duration instead of an integer

File: test/k8sT/Nightly.go [hunk 4]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
-		Expect(err).Should(BeNil())
+		It("Check Kubernetes Example is working correctly", func() {
+			kubectl.Apply(demoPath).ExpectSuccess()
+			_, err := kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).Should(BeNil())
 
-		_, err = kubectl.CiliumPolicyAction(helpers.KubeSystemNamespace, l3Policy, helpers.KubectlApply, 300)
-		Expect(err).Should(BeNil())
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, l3Policy, helpers.KubectlApply, 300)
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Use time.Duration instead of an integer

File: test/k8sT/Nightly.go [hunk 5]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
 
-	It("Check Kubernetes Example is working correctly", func() {
-		var path = "../examples/kubernetes/cilium.yaml"
-		var result bytes.Buffer
-		newCiliumDSName := fmt.Sprintf("cilium_ds_%s.json", helpers.MakeUID())
+	Context("Cilium DS from example", func() {
 
-		objects, err := helpers.DecodeYAMLOrJSON(path)
-		Expect(err).To(BeNil())
+		InstallExampleCilium := func() {
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Add some documentation for this function briefly describing what it does.

File: test/k8sT/Nightly.go [hunk 6]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
+		})
+
+		It("Updating Cilium stable to master", func() {
+			By("Creating some endpoints and L7 policy")
+			kubectl.Apply(demoPath).ExpectSuccess()
+			_, err := kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).Should(BeNil())
+
+			kubectl.Exec(fmt.Sprintf("%s scale deployment %s --replicas=10",
+				helpers.KubectlCmd, helpers.App1)).ExpectSuccess("Cilium DS cannot be deleted")
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: If the command is "scale deployment", why does the message in `ExpectSuccess` say "Cilium DS cannot be deleted"?

File: test/k8sT/Nightly.go [hunk 7]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
+
+			_, err = kubectl.CiliumPolicyAction(
+				helpers.KubeSystemNamespace, l7Policy, helpers.KubectlApply, 300)
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Use time.Duration instead of an integer

File: test/k8sT/Nightly.go [hunk 8]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
+			Expect(err).Should(BeNil(), "cannot import l7 policy")
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: refer to the path of the policy that cannot be imported in this error message

File: test/k8sT/Nightly.go [hunk 9]
```diff
@@ -269,62 +279,164 @@ var _ = Describe("NightlyExamples", func() {
...
+			err = helpers.WithTimeout(
+				waitForUpdateImage,
+				"Cilium Pods are not updating correctly",
+				&helpers.TimeoutConfig{Timeout: 300})
+			Expect(err).To(BeNil(), "Pods are not updating")
+
+			status, err := kubectl.WaitforPods(
+				helpers.KubeSystemNamespace, "-l k8s-app=cilium", 300)
+			Expect(status).Should(BeTrue())
+			Expect(err).Should(BeNil())
```
Comment (ianvernon) [inline test/k8sT/Nightly.go]: Add descriptions to these two `Should` checks.

**Review by ianvernon — DISMISSED** [review]
Mainly doc / log message concerns. Overall I think this is a great addition to our testing framework! While not blocking, I think it would be highly beneficial to have policies imported into Cilium as well that are tested after upgrade - i.e., does upgrade affect policy enforcement?

**Comment by eloycoto** [comment]
test-me-please

**Comment by ianvernon** [comment]
@eloycoto please fix the merge conflicts and I will re-review

**Comment by eloycoto** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**